### DISCUSSION TITLE
Why are UI elements not updating in function?

### DISCUSSION THREAD

**Discussion by AdamW88** [zauberzeug/nicegui#468] [Q&A] [answered]
I am making a simple GUI that lets a user run a Python script and update my GUI elements when the user presses 'Enter'. The basic order of operations is:

1. GUI initializes and is ready for user input
2. User begins to run a test with GUI elements updating to show test is running with a large, green spinner and simple text label
3. Test finishes; GUI elements update to reflect completion

When I replaced the demo function with my actual test script, the test script worked completely as expected, but none of the UI elements updated in the order I expected them to **unless I add in an asyncio.sleep**...Can someone help me understand why?

I am quite new to Python development, so can anyone please let me know what I am doing wrong? A simplified version of what I am doing is below:

```python
from nicegui import ui
from functools import partial
from time import sleep
import asyncio

# Function to invoked when user presses Enter
async def begin_test(sn):
    #Update test_status row to show a green spinner and updated label text
    ui.notify(position = 'center', message = f'Starting test on serial number {sn.value}', type = 'info')
    with test_status:
        current_state.text = ('Testing...')
        ui.spinner(size='lg', color = 'green', thickness = '10')
    ui.update(test_status)
    await asyncio.sleep(2) # Code only seems to work with this asyncio.sleep...why??
    
    #Runs my test script; 100% works as expected
    test_result = await run_test()

    with test_status: #remove the spinner to show test is done
        test_status.remove(1)
    if test_result == True:
        with test_status:
            ui.notify(message='Pass!', position='center', type = 'positive')
    else:
        log_record = 'FAIL'
        with test_status:
            ui.notify(message='Fail!', position='center', type = 'negative')
 
# Personal test script that completely works even with the GUI issues
async def run_test():
    #Runs a bunch of other code and returns a test result
    return ('Fail')

barcode = ui.row()
test_status = ui.row()

with barcode: # Input field for a barcode alphanumeric number
    serial_number = ui.input()
    serial_number.on('keydown.enter', partial(begin_test, (serial_number)))

with test_status: # Shows the current status of the test UI
    current_state = ui.label(text = 'Scan barcode and press enter to begin')

ui.run()
```

**Comment by rodja** [comment]
You write 

> Instead of the expected order of operations above, I see the below:

But there is nothing below this line. Maybe you forgot to add a screenshot?

**Comment by rodja** [comment, marked as answer]
> When I replaced the demo function with my actual test script, the test script worked completely as expected, but none of the UI elements updated in the order I expected them to

Have you looked at the [script executor example](https://github.com/zauberzeug/nicegui/blob/main/examples/script_executor/main.py)? It's vital that all your i/o bound business code is async. Also note that computation heavy tasks should be scheduled into an other process.

**Reply by AdamW88** [reply]
Hi there, thanks for the reply.

> Have you looked at the [script executor example](https://github.com/zauberzeug/nicegui/blob/main/examples/script_executor/main.py)? It's vital that all your i/o bound business code is async. Also note that computation heavy tasks should be scheduled into an other process.

Somehow I did not see this example, thank you I will take a look. Yesterday I found that adding an asyncio.sleep(2) fixed my code completely and made the UI elements update as I expected. I've reflected this change in the example code snippet above. Could you please help me understand why adding this sleep worked? Is it giving the GUI time to refresh the elements that I've requested with the ui.update() call?

**Comment by rodja** [comment]
If you move the `sleep(2)` into the `run_test` function everything should make sense:

```py
async def run_test():
    await asyncio.sleep(2)  # simulates a test that takes 2 seconds in another process or thread
    return ('Fail')
```

So basically your run_test needs to be really taking some time (here simulated by a sleep). Otherwise the interpreter will keep executing each line in `begin_test` until the end. Therefore you are not seeing the spinner at all.
